core/src/cluster_nodes.rs

```diff
@@ -458,27 +460,36 @@ pub fn make_test_cluster<R: Rng>(
     HashMap<Pubkey, u64>, // stakes
     ClusterInfo,
 ) {
+    use solana_gossip::contact_info::ContactInfo;
     let (unstaked_numerator, unstaked_denominator) = unstaked_ratio.unwrap_or((1, 7));
-    let mut nodes: Vec<_> = repeat_with(|| ContactInfo::new_rand(rng, None))
-        .take(num_nodes)
-        .collect();
+    let mut nodes: Vec<_> = repeat_with(|| {
+        let pubkey = solana_sdk::pubkey::new_rand();
+        ContactInfo::new_localhost(&pubkey, /*wallclock:*/ timestamp())
+    })
+    .take(num_nodes)
+    .collect();
     nodes.shuffle(rng);
     let keypair = Arc::new(Keypair::new());
-    nodes[0].id = keypair.pubkey();
+    nodes[0].set_pubkey(keypair.pubkey());
     let this_node = nodes[0].clone();
     let mut stakes: HashMap<Pubkey, u64> = nodes
         .iter()
         .filter_map(|node| {
             if rng.gen_ratio(unstaked_numerator, unstaked_denominator) {
                 None // No stake for some of the nodes.
             } else {
-                Some((node.id, rng.gen_range(0, 20)))
+                Some((*node.pubkey(), rng.gen_range(0, 20)))
             }
         })
         .collect();
     // Add some staked nodes with no contact-info.
     stakes.extend(repeat_with(|| (Pubkey::new_unique(), rng.gen_range(0, 20))).take(100));
     let cluster_info = ClusterInfo::new(this_node, keypair, SocketAddrSpace::Unspecified);
+    let nodes: Vec<_> = nodes
+        .iter()
+        .map(LegacyContactInfo::try_from)
+        .collect::<Result<_, _>>()
+        .unwrap();
```

Comment:
this one

...but mainly due to the multiple locations where `ContactInfo`s are sanitized, both internally and externally, making it difficult to reason about when this might fail

Reply:
ah, dammit. one day rust will get accurate diff context locality annotations